man/mkbinhmmlist.1
'\" t
.\"     Title: 
    mkbinhmmlist
  
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 12/19/2013
.\"    Manual: 
.\"    Source: 
.\"  Language: English
.\"
.TH "MKBINHMMLIST" "1" "12/19/2013" ""
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"

    mkbinhmmlist
   \- convert HMMList file into binary format
.SH "SYNOPSIS"
.HP \w'\fBmkbinhmmlist\fR\ 'u
\fBmkbinhmmlist\fR {hmmdefs_file} {HMMList_file} {output_binhmmlist_file}
.SH "DESCRIPTION"
.PP

\fBmkbinhmmlist\fR
converts a HMMList file to binary format\&. Since the index trees for lookup are also stored in the binary format, it will speed up the startup of Julius, namely when using big HMMList file\&.
.PP
For conversion, HMM definition file
\fIhmmdefs_file\fR
that will be used together at Julius needs to be specified\&. The format of the HMM definition file can be either ascii or Julius binary format\&.
.PP
The output binary file can be used in Julius as the same by "\fB\-hlist\fR"\&. The format wil be auto\-detected by Julius\&.
.PP
The pseudo phone extracted form acoustic model is outputed by version 4\&.2 or later\&. This cause faster starting Julius\&. However, It is required that using it with created acoustic model\&. Also, binhmmlist file created by mkbinhmmlist of this version or later cannot be used in earlier version\&.
.PP

\fBmkbinhmmlist\fR
can read gzipped file\&.
.SH "OPTIONS"
.PP
\fIhmmdefs_file\fR
.RS 4
Acoustic HMM definition file, in HMM ascii format or Julius binary format\&.
.RE
.PP
\fIHMMList_file\fR
.RS 4
Source HMMList file
.RE
.PP
\fIoutput_binhmmlist_file\fR
.RS 4
Output file, will be overwritten if already exist\&.
.RE
.SH "EXAMPLES"
.PP
Convert a HMMList file
logicalTri
into binary format and store to
logicalTri\&.bin:
.sp .if n \{\ .RS 4 .\} .nf % \fBmkbinhmmlist\fR binhmm logicalTri logicalTri\&.bin .fi .if n \{\ .RE .\}
.SH "SEE ALSO"
.PP

\fB julius \fR( 1 )
,
\fB mkbinhmm \fR( 1 )
.SH "COPYRIGHT"
.PP
Copyright (c) 1991\-2013 Kawahara Lab\&., Kyoto University
.PP
Copyright (c) 1997\-2000 Information\-technology Promotion Agency, Japan
.PP
Copyright (c) 2000\-2005 Shikano Lab\&., Nara Institute of Science and Technology
.PP
Copyright (c) 2005\-2013 Julius project team, Nagoya Institute of Technology
.SH "LICENSE"
.PP
The same as Julius\&.


ChangeLog
2008-05-17 13:06  sumomo

	* libjulius/src/: instance.c, m_fusion.c, m_options.c: Now GMM
	  module uses the analysis parameter of the given AM when "-AM_GMM"
	  is not explicitly specified.	Now output error when GMM has
	  different sampling rate from AM.

2008-05-13 15:32  sumomo

	* jcontrol/jcontrol.c: Flush output after each message.

2008-05-13 15:32  sumomo

	* libsent/src/util/readfile.c: Fixed file name prompt to ignore
	  empty input correctly.

2008-05-09 15:00  sumomo

	* libjulius/src/m_usage.c: Grouped the usage messages by the
	  belonging modules.

2008-05-09 14:59  sumomo

	* libjulius/src/m_options.c: New option "-usepower" to use power
	  instead of magnitude in filterbank analysis.

2008-05-09 14:58  sumomo

	* libsent/: include/sent/mfcc.h, src/hmminfo/read_binhmm.c,
	  src/hmminfo/write_binhmm.c, src/wav2mfcc/mfcc-core.c: Added
	  support for USEPOWER in filterbank analysis.

2008-05-09 14:57  sumomo

	* libsent/src/wav2mfcc/para.c: Add support for USEPOWER.  Fixed
	  -htkconf to assume SOURCERATE=625 (16kHz) if not exist.

2008-05-08 14:45  sumomo

	* libjulius/: include/julius/extern.h, src/adin-cut.c,
	  src/m_adin.c: Add tiny value check for frequency.

2008-05-08 14:44  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fixed failure to startup
	  when using multiple sound cards on ALSA.

2008-05-08 14:43  sumomo

	* libsent/src/adin/adin_file.c: Fixed segfault when failed to open
	  audio file.

2008-04-27 22:06  sumomo

	* libjulius/: include/julius/jconf.h, src/default.c, src/m_info.c,
	  src/m_options.c, src/m_usage.c, src/search_bestfirst_main.c: Add
	  an option "-fallback1pass" to use the 1st pass result as final
	  result when the 2nd pass fails.  This was the default on 3.x.

2008-04-27 22:06  sumomo

	* libjulius/src/version.c.in: Update copyright output.

2008-03-28 13:52  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fix ALSA A/D-in to use
	  device other than "default".

2008-03-28 12:57  sumomo

	* julius/record.c: Fixed record sometimes fails.

2008-03-28 12:51  sumomo

	* libjulius/src/adin-cut.c: Moved CALLBACK_ADIN_TRIGGERED to adin
	  thread.

2008-03-25 17:12  sumomo

	* libjulius/src/pass1.c: Fixed GMM rejection not working well on
	  microphone input (enbuged on 4.0.1).

2008-03-21 11:59  sumomo

	* libsent/src/wav2mfcc/para.c: Add warning when HTK Config given by
	  "-htkconf" has unsupported values.

2008-03-20 18:08  sumomo

	* libsent/: include/sent/ngram2.h, include/sent/util.h,
	  src/ngram/ngram_access.c, src/ngram/ngram_compact_context.c,
	  src/ngram/ngram_read_arpa.c, src/ngram/ngram_read_bin.c,
	  src/ngram/ngram_util.c, src/ngram/ngram_write_bin.c,
	  src/util/endian.c: Update N-gram codes to (hopefully) allow 64bit
	  index in the future.

2008-03-20 18:06  sumomo

	* julius/: charconv_win32.c, libjcode/jlib.h, libjcode/libjcode.c:
	  Fixed euc-jp conversion not working on some string.

2008-03-18 02:00  sumomo

	* libjulius/src/: adin-cut.c, recogmain.c: Fixed
	  CALLBACK_EVENT_SPEECH_* call timing when -spsegment/VAD enabled.

2008-03-18 01:59  sumomo

	* libjulius/src/version.c.in, libsent/src/util/confout.c: Fix
	  typos.

2008-03-18 01:58  sumomo

	* julius/main.c, julius/output_stdout.c,
	  libjulius/include/julius/global.h, libjulius/src/callback.c,
	  libjulius/src/m_options.c, libjulius/src/m_usage.c:
	  "-callbackdebug" moved from Julius to JuliusLib.

2008-03-14 22:28  sumomo

	* libsent/src/ngram/ngram_compact_context.c: Fixed a bug that
	  prevents reading big N-gram.

2008-03-12 15:43  tag release_4_0_1

2008-03-12 15:43  sumomo

	* 00readme-ja.txt, 00readme.txt, ChangeLog, LICENSE.txt,
	  Release-ja.txt, Release.txt, libjulius/configure,
	  libjulius/configure.in, libsent/configure, libsent/configure.in,
	  support/build-all.sh: Release of 4.0.1.

2008-03-12 14:48  sumomo

	* libjulius/src/pass1.c: Fixed compilation failure enbuged at last
	  fix...

2008-03-10 11:39  sumomo

	* libjulius/src/: pass1.c, realtime-1stpass.c, recogmain.c: Fix CMN
	  not working well on spsegment/GMM-VAD/decoder-VAD mode.

2008-03-02 02:10  sumomo

	* libsent/: configure, configure.in, include/sent/config.h.in,
	  src/adin/adin_mic_linux_alsa.c: Fixed ALSA recording to work.
	  Many codes are updated.

2008-02-21 13:10  sumomo

	* libjulius/src/beam.c: Word recognition results are not correctly
	  sorted when using "-output N".

2008-02-18 23:35  sumomo

	* julius/module.c: Added a module command "GRAMINFO" to return
	  grammar info.

2008-02-18 23:32  sumomo

	* libjulius/src/m_fusion.c: Now -spsegment detects error when pause
	  word is not defined in dictionary.

2008-02-15 03:26  sumomo

	* libsent/src/phmm/mkwhmm.c: Fixed bug of immediate stop when no sp
	  model in multi-path mode.

2008-02-15 03:25  sumomo

	* libjulius/src/: m_info.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c, word_align.c: Fixed hungup when sp model
	  not exist on AM in multipath.

2008-02-14 13:44  sumomo

	* libjulius/: include/julius/extern.h, src/multi-gram.c: Changes
	  for better grammar handling: - multigram_add() returns the newly
	  assigned grammar id.	- a function is added to get grammar id
	  from name string.  - modified STAT message to output grammar name
	  together with id.

2008-02-14 13:41  sumomo

	* julius/module.c: Extended ACTIVATEGRAM, DEACTIVATEGRAM, DELGRAM
	  to specify grammar by name.

2008-02-13 17:49  sumomo

	* libjulius/include/julius/extern.h: Updated for some function
	  changes.

2008-02-13 17:48  sumomo

	* libjulius/src/: outprob_style.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c: Global variables and work area has been
	  moved to RecogProcess and StackDecode.

2008-02-13 17:47  sumomo

	* libjulius/src/wchmm.c: qsort function switched to qsort_reentrant
	  in libsent for modularity.

2008-02-13 17:46  sumomo

	* libjulius/src/confnet.c: Updated for being re-entrant.

2008-02-13 17:45  sumomo

	* libjulius/src/: m_jconf.c, m_options.c, ngram_decode.c,

mkbinhmm/mkbinhmmlist.c
/*
 * Copyright (c) 2003-2013 Kawahara Lab., Kyoto University 
 * Copyright (c) 2003-2005 Shikano Lab., Nara Institute of Science and Technology
 * Copyright (c) 2005-2013 Julius project team, Nagoya Institute of Technology
 * All rights reserved
 */

/* mkbinhmmlist --- read in ascii hmmlist file and write in binary format */

/* $Id: mkbinhmmlist.c,v 1.5 2013/06/20 17:14:27 sumomo Exp $ */

#include <sent/stddefs.h>
#include <sent/htk_hmm.h>


HTK_HMM_INFO *hmminfo;
Value para, para_htk;


static void
usage(char *s)
{
  printf("mkbinhmmlist: convert HMMList file to binary format for Julius\n");
  printf("usage: %s hmmdefs hmmlist binhmmlist\n", s);
  printf("\nLibrary configuration: ");
  confout_version(stdout);
  confout_am(stdout);
  printf("\n");
}


int
main(int argc, char *argv[])
{
  FILE *fp;
  char *hmmdefs_file;
  char *hmmlist_file;
  char *outfile;
  int i;

  hmmdefs_file = hmmlist_file = outfile = NULL;
  for(i=1;i<argc;i++) {
    if (hmmdefs_file == NULL) {
      hmmdefs_file = argv[i];
    } else if (hmmlist_file == NULL) {
      hmmlist_file = argv[i];
    } else if (outfile == NULL) {
      outfile = argv[i];
    } else {
      usage(argv[0]);
      return -1;
    }
  }
  if (hmmdefs_file == NULL || hmmlist_file == NULL || outfile == NULL) {
    usage(argv[0]);
    return -1;
  }

  hmminfo = hmminfo_new();

  printf("---- reading hmmdefs ----\n");
  printf("filename: %s\n", hmmdefs_file);

  /* read hmmdef file */
  undef_para(&para);
  if (init_hmminfo(hmminfo, hmmdefs_file, hmmlist_file, &para) == FALSE) {
    fprintf(stderr, "--- terminated\n");
    return -1;
  }

  if (hmminfo->is_triphone) {
    fprintf(stderr, "making pseudo bi/mono-phone for IW-triphone\n");
    if (make_cdset(hmminfo) == FALSE) {
      fprintf(stderr, "ERROR: m_fusion: failed to make context-dependent state set\n");
      return -1;
    }
  }

  printf("\n------------------------------------------------------------\n");
  print_hmmdef_info(stdout, hmminfo);
  printf("\n");

  printf("------------------------------------------------------------\n");

  printf("---- writing logical-to-physical mapping and pseudo phone info ----\n");
  printf("filename: %s\n", outfile);

  if ((fp = fopen_writefile(outfile)) == NULL) {
    fprintf(stderr, "failed to open %s for writing\n", outfile);
    return -1;
  }
  if (save_hmmlist_bin(fp, hmminfo) == FALSE) {
    fprintf(stderr, "failed to write to %s\n", outfile);
    return -1;
  }
  if (fclose_writefile(fp) != 0) {
    fprintf(stderr, "failed to close %s\n", outfile);
    return -1;
  }

  printf("\n");
  printf("binary HMMList and pseudo phone definitions are written to \"%s\"\n", outfile);

  return 0;
}
